ENH improve error message in _get_response function

shinehide opened the pull request:
Closes #20577

Improve error message raised in `_get_response` when an estimator other than a binary classifier is given.

shinehide committed:
fix issue #20577

It could be improved, e.g., thatt may be confusing to get "KMeans should be a binary classifier",  https://github.com/scikit-learn/scikit-learn/issues/20577

glemaitre commented:
You need to change the tests that are failing since we change the error message.
In addition, you need to add an entry in the what's new as well.

lucyleeow commented:
Could you also look at addressing this: https://github.com/scikit-learn/scikit-learn/issues/20577#issuecomment-886539332 (if the parameter is common amongst the function uses, it would be nice to specify the parameter this warning corresponds to)

shinehide committed:
update test message

Update test message and doc, make test message same as error message

shinehide commented:
As your suggestion, i update test message and doc. But for this comment  #20577 (comment) , I understand that specific parameters need to be printed to facilitate users to clearly understand the error. But I don't know which parameter it is. Maybe you can guide me to complete this further work. Thank you very much.

glemaitre commented:
The tests are failing.

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -134,6 +134,7 @@ Changelog
 ...................
 
 - |Fix| :func:`config_context` is now threadsafe. :pr:`18736` by `Thomas Fan`_.
+- |Fix| :func:`_get_response` make error message clear. :pr:`20618` by `shinehide`_.
```
We don't need an entry in what's new here. We can add the label in the PR to discard one of the CI

shinehide commented:
As you said 'The tests are failing.'.  Does the test mean ‘sklearn/metrics/_plot/tests/test_plot_precision_recall.py’  and 'sklearn/metrics/_plot/tests/test_plot_curve_common.py' ?

shinehide commented:
As you said 'The tests are failing.'.  Does the test mean ‘sklearn/metrics/_plot/tests/test_plot_precision_recall.py’  and 'sklearn/metrics/_plot/tests/test_plot_curve_common.py' ?

lucyleeow commented on sklearn/metrics/_plot/base.py:
```diff
@@ -82,7 +82,7 @@ def _get_response(X, estimator, response_method, pos_label=None):
         the metrics.
     """
-    classification_error = "{} should be a binary classifier".format(
+    classification_error = "Expected a binary classifier, but got {}".format(
```
Could you use a f-string here, it's less verbose and helps readability (ref: https://github.com/scikit-learn/scikit-learn/issues/17327)

shinehide replied:
Thank you , Lucy

lucyleeow commented:
Some notes on running tests locally here: https://scikit-learn.org/stable/developers/contributing.html#pull-request-checklist

To check which tests are failing you can click into 'details' part of the CIs, e.g.: https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=30902&view=logs&j=13650d7c-49e8-54f3-0598-c3480d1c1e4f&t=97f2162d-a0c4-54dc-9fce-2fceef7172e3&l=859

>I understand that specific parameters need to be printed to facilitate users to clearly understand the error. But I don't know which parameter it is. Maybe you can guide me to complete this further work. Thank you very much.

Take a look at all the places `_get_response` is used to see if the parameter the message is referring to is common. It seems that everywhere it is used `estimator` is the parameter the error refers to. So the message could be something like: `Expected 'estimator' to be a binary classifier, but got {}`

shinehide committed:
update to f-string

I checked all the places where the '_get_response'  was called and the parameters used.   Seems that everywhere it is used estimator. So change the error message.

glemaitre commented on sklearn/metrics/_plot/base.py:
```diff
@@ -82,9 +82,7 @@ def _get_response(X, estimator, response_method, pos_label=None):
-    classification_error = "Expected a binary classifier, but got {}".format(
-        estimator.__class__.__name__
-    )
+    classification_error = f"Expected 'estimator' to be a binary classifier, but got {estimator.__class__.__name__}"
```
This line is surely too long and will be caught by triggering an error on the CI

shinehide replied:
I split it and added a new line,   hope my line feed method meets the code specification.

shinehide committed:
Update base.py

shinehide commented:
I checked all the places where the '_get_response'  was called and the parameters used.   Seems that everywhere it is used estimator. So change the error message.
If it still doesn't pass build & test, I'll fix it the weekend.

shinehide committed:
Update v1.0.rst

shinehide committed:
The lines too long

the lines too long to pass CI

shinehide commented:
`2021-07-29T15:20:12.1524459Z ========================== Starting Command Output ===========================
2021-07-29T15:20:12.1542741Z [command]/usr/bin/bash --noprofile --norc /home/vsts/work/_temp/1b2faa6b-821f-4936-b25e-c39b74f8c1e6.sh
2021-07-29T15:20:34.5493842Z would reformat sklearn/metrics/_plot/base.py
2021-07-29T15:20:34.9080192Z would reformat sklearn/metrics/_plot/tests/test_plot_curve_common.py
2021-07-29T15:20:35.1359358Z would reformat sklearn/metrics/_plot/tests/test_plot_precision_recall.py
2021-07-29T15:20:52.9519261Z Oh no! 💥 💔 💥
2021-07-29T15:20:52.9520610Z 3 files would be reformatted, 577 files would be left unchanged.
2021-07-29T15:20:52.9814726Z ##[error]Bash exited with code '1'.
2021-07-29T15:20:52.9831917Z ##[section]Finishing: Run black`

I checked the log, but I didn't find any useful information... Do you have any suggestions. Thanks

glemaitre commented:
You need to run black on the changed file. The best would be to do the optional step 9. from the the contributing guide: https://scikit-learn.org/stable/developers/contributing.html#how-to-contribute

It will install black and make the different checks before committing. All style issues will be given before making a push.

shinehide committed:
Format Code

Format code with black

shinehide committed:
Update test_plot_precision_recall.py

shinehide commented:
@glemaitre 
Hi glemaitre ,
Is there anything I need to do for this PR, or just wait for merge?

glemaitre reviewed (approved):
LGTM. @lucyleeow  do you want to have a look?

lucyleeow reviewed (approved):
LGTM! Thanks  @shinehide @glemaitre !

glemaitre commented:
Thanks @shinehide @lucyleeow. Merging.
